src/examples/3d_fluid_sim/fluid_sim.cpp

```diff
@@ -0,0 +1,198 @@
+#include "fluid_sim.hpp"
+
+#define simIterations 250
+#define DECLARE_VARIABLES \
+    wash::add_variable("deltaTime",  (1.0/60.0) / 3.0 * 0.9 ); \
+    wash::add_variable("gravity", -10.0); \
+    wash::add_variable("collisionDamping", 0.95); \
+    wash::add_variable("smoothingRadius", 0.2); \
+    wash::add_variable("targetDensity", 630.0); \
+    wash::add_variable("pressureMultiplier", 288.0); \
+    wash::add_variable("nearPressureMultiplier", 2.25); \
+    wash::add_variable("viscosityStrength", 0.08);
+#define numParticles 42875
+#define numParticlesPerAxis 35
+#define boundsSize wash::Vec3D {9.1, 9.3, 9.1}  // Bound Size of simulation
+#define centre wash::Vec3D {0.0, -0.47, 0.0}
+#define initialVel wash::Vec3D { 0.0, 0.0, 0.0 }
+#define size 3.7
+#define jitterStrength 0.035
+
+void spawn_particles() {
+
+    for (int x = 0; x < numParticlesPerAxis; x++) {
+        for (int y = 0; y < numParticlesPerAxis; y++) {
+            for (int z = 0; z < numParticlesPerAxis; z++) {
+
+                double tx = x / (numParticlesPerAxis - 1.0);
+                double ty = y / (numParticlesPerAxis - 1.0);
+                double tz = z / (numParticlesPerAxis - 1.0);
+
+                double px = (tx - 0.5) * size + centre.at(0);
+                double py = (ty - 0.5) * size + centre.at(1);
+                double pz = (tz - 0.5) * size + centre.at(2);
+
+                wash::Vec3D jitter = randomSpherePoint() * jitterStrength;
+                auto pos = wash::Vec3D{ px, py, pz } + jitter;
+
+                auto& p = wash::create_particle(0.0, 1.0, wash::get_variable("smoothingRadius"), pos, initialVel);
+                p.set_force_vector("predictedCoordinates", p.get_pos());
+            }
+        }
+    }
+
+}
+
+void external_forces(wash::Particle& particle) {
+    particle.set_vel(particle.get_vel() + wash::Vec3D { 0.0, wash::get_variable("gravity"), 0.0 } * wash::get_variable("deltaTime"));
+
+    particle.set_force_vector("predictedCoordinates", particle.get_pos() + particle.get_vel() * 1.0/120.0);
+}
+
+void density(wash::Particle& particle, const std::vector<wash::Particle>& neighbours) {
+    auto pos = particle.get_force_vector("predictedCoordinates");
+    double density = 1.0;
+    double near_density = 1.0;
+
+    for (auto& p : neighbours) {
+        auto npos = p.get_force_vector("predictedCoordinates");
+        auto dst = (pos - npos).magnitude();
+
+        density += density_kernel(dst, wash::get_variable("smoothingRadius"));
+        near_density += near_density_kernel(dst, wash::get_variable("smoothingRadius"));
+    }
+
+    particle.set_density(density);
+    particle.set_force_scalar("nearDensity", near_density);
+}
+
+void pressure(wash::Particle& particle, const std::vector<wash::Particle>& neighbours) {
+    auto pos = particle.get_force_vector("predictedCoordinates");
```

Comment:
For these ones that aren't changing over the course of the function, you can mark them as `const auto`

Reply:
See previous comment, transform to constexpr globals made.